codexdht/private/eth/p2p/discoveryv5/protocol.nim

```diff
@@ -499,15 +501,33 @@ proc waitNodes(d: Protocol, fromNode: Node, reqId: RequestId):
   ## on that, more replies will be awaited.
   ## If one reply is lost here (timed out), others are ignored too.
   ## Same counts for out of order receival.
+  let startTime = Moment.now()
   var op = await d.waitMessage(fromNode, reqId)
   if op.isSome:
     if op.get.kind == MessageKind.nodes:
       var res = op.get.nodes.sprs
-      let total = op.get.nodes.total
+      let
+        total = op.get.nodes.total
+        firstTime = Moment.now()
+        rtt = firstTime - startTime
+      # trace "nodes RTT:", rtt, node = fromNode
+      fromNode.registerRtt(rtt)
       for i in 1 ..< total:
         op = await d.waitMessage(fromNode, reqId)
         if op.isSome and op.get.kind == MessageKind.nodes:
           res.add(op.get.nodes.sprs)
+          # Estimate bandwidth based on UDP packet train received, assuming these were
+          # released fast and spaced in time by bandwidth bottleneck. This is just a rough
+          # packet-pair based estimate, far from being perfect.
+          # TODO: get message size from lower layer for better bandwidth estimate
+          # TODO: get better reception timestamp from lower layers
+          let
+            deltaT = Moment.now() - firstTime
+            bwBps = 500.0 * 8.0 / (deltaT.nanoseconds.float / i.float / 1e9)
```

Comment:
Not sure I get this calculation. Why 500 * 8? Is this a 500 byte packet?

Reply:
As mentioned in the TODO above, I don't have size information (yet). This is around 500 bytes in most cases, but of course surfacing the size would be better. I needs a follow-up (hence the TODO).

Reply:
We can surface the size information (and also a better receive timestamp) relatively easily, but I would leave that for a follow-up PR. For now let's just accept that this bandwidth estimate is only a rough estimate, and deal with this if/when we start to use this value for something.